File: BasicSampleJava/app/src/main/java/com/regula/facesample/adapter/CategoryAdapter.java
package com.regula.facesample.adapter;

import android.content.Context;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.TextView;

import androidx.annotation.NonNull;
import androidx.recyclerview.widget.RecyclerView;

import com.regula.facesample.R;
import com.regula.facesample.data.CategoryDataProvider;
import com.regula.facesample.items.CategoryItem;
import com.regula.facesample.items.ICategoryItem;

import java.lang.ref.WeakReference;
import java.util.List;

/**
 * Created by Sergey Yakimchik on 20.08.21.
 * Copyright (c) 2021 Regula. All rights reserved.
 */

public class CategoryAdapter extends RecyclerView.Adapter<RecyclerView.ViewHolder> {

    public static final int SECTION_VIEW = 0;
    public static final int CONTENT_VIEW = 1;

    private final List<ICategoryItem> items;
    private final WeakReference<Context> contextRef;

    public CategoryAdapter(@NonNull Context context, CategoryDataProvider dataProvider) {
        this.items = dataProvider.getCategoryItems();
        this.contextRef = new WeakReference<>(context);
    }

    @NonNull
    @Override
    public RecyclerView.ViewHolder onCreateViewHolder(@NonNull ViewGroup parent, int viewType) {

        Context context = contextRef.get();
        if (viewType == SECTION_VIEW) {
            return new SectionHeaderViewHolder(LayoutInflater.from(parent.getContext()).inflate(R.layout.item_header_title, parent, false));
        }
        return new ItemViewHolder(LayoutInflater.from(parent.getContext()).inflate(R.layout.item_element, parent, false));
    }

    @Override
    public int getItemViewType(int position) {
        if (items.get(position).isHeader()) {
            return SECTION_VIEW;
        } else {
            return CONTENT_VIEW;
        }
    }

    @Override
    public void onBindViewHolder(@NonNull final RecyclerView.ViewHolder holder, int position) {

        Context context = contextRef.get();
        if (context == null) {
            return;
        }

        ICategoryItem item = items.get(position);

        if (SECTION_VIEW == getItemViewType(position)) {

            SectionHeaderViewHolder sectionHeaderViewHolder = (SectionHeaderViewHolder) holder;
            sectionHeaderViewHolder.headerTitleTextView.setText(item.getTitle());
            return;
        }

        ItemViewHolder itemViewHolder = (ItemViewHolder) holder;
        CategoryItem categoryItem = (CategoryItem) item;
        itemViewHolder.titleTv.setText(categoryItem.getTitle());
        itemViewHolder.descriptionTv.setText(categoryItem.getDescription());
        itemViewHolder.itemView.setOnClickListener(v -> categoryItem.onItemSelected(contextRef.get()));
    }


    @Override
    public int getItemCount() {
        return items.size();
    }

    public static class ItemViewHolder extends RecyclerView.ViewHolder {

        public TextView titleTv, descriptionTv;

        public ItemViewHolder(View itemView) {
            super(itemView);
            titleTv = itemView.findViewById(R.id.titleTv);
            descriptionTv = itemView.findViewById(R.id.descriptionTv);
        }
    }

    public static class SectionHeaderViewHolder extends RecyclerView.ViewHolder {
        TextView headerTitleTextView;

        public SectionHeaderViewHolder(View itemView) {
            super(itemView);
            headerTitleTextView = (TextView) itemView.findViewById(R.id.headerTitleTextView);
        }
    }
}

File: FaceSampleKotlin/app/src/main/java/com/regula/facesamplekotlin/CategoryAdapter.kt
package com.regula.facesamplekotlin

import android.content.Context
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.TextView
import androidx.recyclerview.widget.RecyclerView
import com.regula.facesamplekotlin.category.CategoryItem
import com.regula.facesamplekotlin.category.ICategoryItem
import com.regula.facesamplekotlin.CategoryAdapter.Companion.VIEW_ITEM as VIEW_ITEM1


class CategoryAdapter(private val context: Context, var list: List<ICategoryItem>) :
    RecyclerView.Adapter<RecyclerView.ViewHolder>() {

    companion object {
        const val VIEW_HEADER = 1
        const val VIEW_ITEM = 2
    }

    private inner class HeaderViewHolder(itemView: View) :
        RecyclerView.ViewHolder(itemView) {
        var title: TextView = itemView.findViewById(R.id.headerTitleTextView)
        fun bind(position: Int) {
            val item = list[position]
            title.setText(item.getTitle())
        }
    }

    private inner class ItemViewHolder(itemView: View) :
        RecyclerView.ViewHolder(itemView) {
        var title: TextView = itemView.findViewById(R.id.titleTv)
        var description: TextView = itemView.findViewById(R.id.descriptionTv)
        fun bind(position: Int) {
            val item = list[position] as CategoryItem
            title.setText(item.getTitle())
            item.description?.let { description.setText(it) }

            itemView.setOnClickListener { v: View? ->
                item.onItemSelected(
                    context
                )
            }
        }
    }

    override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): RecyclerView.ViewHolder {
        if (viewType == VIEW_HEADER) {
            return HeaderViewHolder(
                LayoutInflater.from(context).inflate(R.layout.item_header_title, parent, false)
            )
        }
        return ItemViewHolder(
            LayoutInflater.from(context).inflate(R.layout.item_element, parent, false)
        )
    }

    override fun getItemCount(): Int {
        return list.size
    }

    override fun onBindViewHolder(holder: RecyclerView.ViewHolder, position: Int) {
        if (list.get(position).isHeader()) {
            (holder as HeaderViewHolder).bind(position)
        } else {
            (holder as ItemViewHolder).bind(position)
        }
    }

    override fun getItemViewType(position: Int): Int {
        if (list.get(position).isHeader())
            return VIEW_HEADER;
        return VIEW_ITEM1;
    }
}
